### DISCUSSION TITLE
Recursive type alias with variadic generics

### DISCUSSION THREAD

**Discussion by sylee957** [microsoft/pyright#4731] [Q&A]
```python3
from typing import Generic, TypeVar, TypeVarTuple, Unpack, TypeAlias
from dataclasses import dataclass


Ts = TypeVarTuple("Ts")
T = TypeVar("T")


@dataclass(frozen=True, init=False)
class Tree(Generic[*Ts]):
    args: tuple[*Ts]

    def __init__(self, *args: Unpack[Ts]):
        super().__setattr__("args", args)


class A(Tree[*tuple[T, ...]]):
    pass


class B(Tree[*Ts]):
    pass


Foo: TypeAlias = T | A["Foo[T]"]
Bar: TypeAlias = T | B[*tuple["Bar[T]", ...]]
```

Type inference for `A` works 

```python3
x: Foo[int]
x = A(1, 2, 3)
x = A(A(1, 2), 3)
x = A(1, A(2, 3))
```

However, the type inference for `B` does not work for nested cases
```python3
y: Bar[int]
y = B(1, 2, 3)
y = B(B(1, 2), 3)
y = B(1, B(2, 3))
```

Although the typing should be correct
```python
y = B[*tuple[Bar[int], ...]](B(1, 2), 3)
y = B[*tuple[Bar[int], ...]](1, B(2, 3))
y = B(1, B(2, B[*tuple[Bar[int], ...]](3, 4)))
```

Maybe it comes with the technical reason that type inference is not possible for that,
but I would have question that the case with `B` is an issue or should be a pattern to avoid.

I am using pyright via VS Code Pylance extension, with `strict` enabled
(`Pylance language server 2023.3.10 (pyright 773a7cde) starting`)

**Comment by erictraut** [comment]
You're using two of the more complicated and fragile features of the type system — recursive type aliases and variadic generics. The current implementation doesn't allow the two to be combined. Recursive type aliases require the use of bidirectional type inferences, and unpacked TypeVarTuple parameters don't currently support this. Adding this functionality would add quite a bit of complexity to already-complex code, so I'm reluctant to do so unless there's a compelling use case.
